### PR TITLE
✨ v1a2 provider & api changes

### PR DESCRIPTION
**Pull request by bryanv**
Snapshot of the current v1a2 work. This overall effort still has a lot of things to sort out but this works well enough to create working TKGs, with everything else will using v1a1 so we have a baseline of version conversion and functionality.

```release-note
NONE
```

### CODE REVIEW

File: pkg/vmprovider/providers/vsphere2/network/network.go [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+
+	netIf, err = waitForReadyNetworkInterface(vmCtx, client, netIf.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return netIfToResult(vimClient, netIf), nil
+}
+
+func netIfToResult(
```
Comment (sreyasn) [inline pkg/vmprovider/providers/vsphere2/network/network.go:297]: How about netOpNetIfToResult() and ncpNetIfToResult() for the other one below?.

File: pkg/vmprovider/providers/vsphere2/network/network.go [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+	}
+
+	return netIf, nil
+}
+
+// NCPCRName returns the name to be used for the NCP VirtualNetworkInterface CR.
+func NCPCRName(vmName, networkName, interfaceName string, isV1A1 bool) string {
+	var name string
+
+	if isV1A1 {
```
Comment (sreyasn) [inline pkg/vmprovider/providers/vsphere2/network/network.go:377]: Is this referring to the vmop v1a1/v1a2?. If so, should we pass in the FSS val here to NCPCRName(..,.., lib.Isv1a2enabled()) in line 411.

File: pkg/vmprovider/providers/vsphere2/network/network.go [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+	if kind := interfaceSpec.Network.Kind; kind != "" && kind != "Network" {
+		return nil, fmt.Errorf("network kind %q is not supported for VDS", kind)
+	}
+
+	// If empty, NetOP will try to select a namespace default.
+	networkName := interfaceSpec.Network.Name
+
+	netIf := &netopv1alpha1.NetworkInterface{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      NetOPCRName(vmCtx.VM.Name, networkName, interfaceSpec.Name, false),
```
Comment (sreyasn) [inline pkg/vmprovider/providers/vsphere2/network/network.go:268]: Should we pass in the FSS val for v1a2 here to the last argument?. or Is it referring to a different v1a1.

File: pkg/vmprovider/providers/vsphere2/network/network.go [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+
+	vnetIf, err = waitForReadyNCPNetworkInterface(vmCtx, client, vnetIf.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	return vnetIfToResult(vmCtx, vimClient, clusterMoRef, vnetIf)
+}
+
+func vnetIfToResult(
```
Comment (sreyasn) [inline pkg/vmprovider/providers/vsphere2/network/network.go:437]: maybe ncpNetIfToResult() ?

File: pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap_cloudinit.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+				if userdata != "" {
+					break
+				}
+			}
+		}
+
+		// NOTE: The old code didn't error out if userdata wasn't found, so keep going.
+
+	} else {
+		return nil, nil, fmt.Errorf("TODO: inlined CloudConfig")
```
Comment (sreyasn) [inline pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap_cloudinit.go:68]: How do plan on supporting this?. 
do we need a JIRA/issue to track this somewhere?.

**Comment by sreyasn** [comment]
Also JFYI @bryanv , Keep an eye out on the [BackupVirtualMachine()](https://github.com/vmware-tanzu/vm-operator/blob/87313a56d3718b670ef6235560e3be33943e5d15/pkg/vmprovider/interface_a2.go#L24) method added to the v1a2 interface. Without the provider for this the cayman_vm-operator build will fail.

**Review by aruneshpa — APPROVED** [review]

File: pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap_cloudinit.go (hunk 5 above)
Comment (bryanv) [reply pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap_cloudinit.go:68]: Yes there is a series of lower priority BL items for these. My immediate goal is to get v1a2 to v1a1 parity, and then worry more about new functionality like the inline bootstrap configs here.

File: pkg/vmprovider/providers/vsphere2/network/network.go (hunk 2 above)
Comment (bryanv) [reply pkg/vmprovider/providers/vsphere2/network/network.go:376]: No. This is referring to if the VM was a prior a v1a1 VM which means we'll netIF CRDs with the prior naming convention. The existing v1a1 naming convention was ad-hoc and limiting (like only allow one per network) that I want to get away from, especially now that we have a interface name in the spec which was a prototype oversight. If the VM already has netIF CRDs we need to continue to use those and not create new ones.

This is part of the larger issue to ensure existing VMs continue to be reconciled properly post the switch to v1a2.

File: pkg/vmprovider/providers/vsphere2/network/network.go (hunk 1 above)
Comment (bryanv) [reply pkg/vmprovider/providers/vsphere2/network/network.go:296]: Sure

File: pkg/vmprovider/providers/vsphere2/network/network.go (hunk 4 above)
Comment (bryanv) [reply pkg/vmprovider/providers/vsphere2/network/network.go]: Sure

**Review by sreyasn — APPROVED** [review]
Thank you for this change, BV..

**Pull request merged**